core/dnsserver/server.go

```diff
@@ -314,9 +320,32 @@ func DefaultErrorFunc(w dns.ResponseWriter, r *dns.Msg, rc int) {
 	w.WriteMsg(answer)
 }
 
+// depthCheck increments the loop counter in the context, and returns an error if
+// the counter exceeds the max number of re-entries
+func depthCheck(ctx context.Context) (context.Context, error) {
+	// Loop counter for self directed lookups
+	loop := ctx.Value(loopKey)
+	if loop == nil {
+		ctx = context.WithValue(ctx, loopKey, 0)
+	} else {
+		iloop := loop.(int) + 1
+		if iloop > maxreentries {
+			return ctx, fmt.Errorf("too deep")
+		}
+		ctx = context.WithValue(ctx, loopKey, iloop)
+	}
+	return ctx, nil
+}
+
+type key int
+
 const (
-	tcp = 0
-	udp = 1
+	tcp          = 0
+	udp          = 1
+	maxreentries = 10
+	// ServerKey is the context key for the current server
+	ServerKey key = 0
```

Comment:
must it be exported?
the full var now reads `dnsserver.ServerKey` think just `Key` read better?

Reply:
I have seen others use a typed, empty struct as the key. That way there is no collision with someone else using the context. See https://github.com/opentracing/opentracing-go/blob/30dda9350627161ff15581c0bdc504e32ec9a536/gocontext.go#L5